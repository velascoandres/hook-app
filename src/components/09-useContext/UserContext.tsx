import { createContext } from 'react';

export type UserContext = {
    user: IUser;
    setUser: (user: IUser) => void;
}

export interface IUser {
    id?: number;
    name?: string;
    email?: string;
}

const emptyUser = {
    id: 0,
    name: '',
    email: '',
};

const defaultSetUser = (user: IUser) => { };

export const UserContext = createContext<UserContext>({ user: emptyUser, setUser: defaultSetUser });