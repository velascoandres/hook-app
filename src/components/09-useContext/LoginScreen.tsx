import React, { useContext } from 'react'
import { IUser, UserContext } from './UserContext'

export const LoginScreen = () => {

    // 1. Obtener referencia al context
    // 2. setUser

    const {setUser} = useContext(UserContext);
    
    
    const nuevoUsuario: IUser = {
        email: 'nuevo@gmail.com',
        id: 2,
        name: 'Pepe'
    };


    return (
        <div>
            <h1>Login Screen</h1>
            <hr/>

            <button
                className="btn btn-primary"
                onClick={() => setUser(nuevoUsuario) }
            >
                Establecer Usuario
            </button>
        </div>
    )
}
