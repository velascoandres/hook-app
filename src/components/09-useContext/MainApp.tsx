import { useState } from 'react';
import { AppRouter } from './AppRouter';
import { IUser, UserContext } from './UserContext';


export const MainApp = () => {

    const initialUser: IUser = {
        id: 1234,
        name: 'Andres',
        email: 'correo@mail.com'
    };

    const [user, setUser] = useState<IUser>(initialUser);


    const userContextValue = {user, setUser};


    return (
        <UserContext.Provider value={userContextValue}>
            <AppRouter />
        </UserContext.Provider>
    );
}

