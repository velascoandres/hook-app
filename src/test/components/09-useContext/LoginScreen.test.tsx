import { act } from "@testing-library/react";
import { mount } from "enzyme"
import { LoginScreen } from "../../../components/09-useContext/LoginScreen";
import { IUser, UserContext } from "../../../components/09-useContext/UserContext";

describe('Pruebas con <LoginScreen />', () => {

    const user = {
        name: 'Fernando',
        email: 'fernando@gmail.com'
    };

    const nuevoUsuario: IUser = {
        email: 'nuevo@gmail.com',
        id: 2,
        name: 'Pepe'
    };


    const setUser = jest.fn((user: IUser) => { });
    const wrapper = mount(
        <UserContext.Provider
            value={{ user, setUser }}
        >
            <LoginScreen />
        </UserContext.Provider>
    );

    test('debe de mostrarse correctamente', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('debe de ejecutar el setUser con el argumento esperado', () => {
        act(
            () => {
                wrapper.find('button').simulate('click');
            }
        );
        expect(setUser).toHaveBeenLastCalledWith(nuevoUsuario);
    })

})
