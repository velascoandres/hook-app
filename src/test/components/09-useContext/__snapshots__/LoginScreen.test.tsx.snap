// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas con <LoginScreen /> debe de mostrarse correctamente 1`] = `
<div>
  <h1>
    Login Screen
  </h1>
  <hr />
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Establecer Usuario
  </button>
</div>
`;
